protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+simultaneously. Bundling features in this way simplifies coordination but
+limits flexibility, such as enabling individual features for testing or
+delaying specific ones without forking the config.
+
+In the past, contributors of the OP Stack have had issues surrounding
+hardfork features. For example, long before the Isthmus hardfork, the
+Operator Fee Feature was [proposed][proposed] and approved. The feature
+was considered to be a part of the upcoming Isthmus hardfork. Rust
+implementations of OP Stack components went ahead and implemented the
+feature. For months, the feature sat idle. After it was eventually merged,
+there were issues in the execution client changes for the feature. Removing
+implementations of the feature was considered, but the effort to remove it
+was not seen as worthwhile. After a while, the feature was finally
+complete. It effectively alone delayed the hardfork by a month or so.
+
+This could have been avoided by being able to implement the feature using
+a feature-specific toggle in the `RollupConfig`. The feature toggle would
+be toggled on in the [`superchain-registry`][scr] once the feature is ready.
+In this way, features could be implemented and left in client software
+without being connected to hardfork. The key part: *no changes would
+need to be made to client software to enable the feature*, only enabling
+the feature inside the [`superchain-registry`].
+
+
+# Problem Statement
+
+Bundling features within hardforks limits flexibility:
+
+- All features in a hardfork activate at once, preventing selective enablement
+  for testing or debugging.
+- Custom chains or testnets cannot easily remap or delay individual features
+  without forking entire client implementations.
+- CLI overrides are limited to hardfork timestamps, not per-feature adjustments.
+- The set of features for a given hardfork must be determined prior to
+  implementation, locking and committing to scope.
+
+This rigidity hinders iteration, increases deployment risks, and complicates
+superchain governance.
+
+
+# Current Approach
+
+- Hardforks are defined by named timestamps in the `RollupConfig`
+  (e.g., `fjordTime: uint64`).
+- Node logic checks `timestamp >= hardforkTime` to enable the all features
+  in the hardfork.
+- Features are implicitly tied to their hardfork; no per-feature granularity.
+- CLI overrides exist for hardfork timestamps (e.g., `--override.fjord=123456`
+  in `op-node`).
+    - Overrides in the `op-node` exist [here][op-overrides]. In the `kona-node`,
+      [here][kona-overrides].
+
+
+# Proposed Solution
+
+Decouple features from hardforks by modifying `RollupConfig`s to include a
+mapping from feature names to hardfork names (e.g., a `map[Feature]Fork`
+where the key is feature like "operatorFee" and value is a hardfork like "isthmus").
+If a feature is absent from the map, it remains inactive.
+
+### Key Changes
+
+- **Superchain Registry Updates:** Add a mapping from features to their hardfork
```

Comment:
Do we really need to update the Superchain registry here? I think it would be much simpler to hardcode the hardfork activations in the clients themselves once we agree on a scope of an upgrade. The code deduplication we get from this doesn't seem worth it, and different clients may implement features differently (e.g. a Reth client may have different feature flag names from a Nethermind client), or have a completely different set of feature flags.

Reply:
It seems like adding this to the superchain registry makes sense _if_ the intent is to keep features decoupled from hard forks permanently - ie one chain could have operator fee as part of Isthmus, another activates it as part of Jovian and another never activates it.

Whereas if the decoupling only exists during development of the hard fork and we then when the hard fork scope is finalized and everything is confirmed as ready to ship, we remove the individual feature toggles, locking in what is or isn't part of Isthmus, then I'd say we don't need it in the superchain registry since most devnets aren't in the superchain registry and we can keep things simpler.